src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,294 @@
+ * CAmount not_input_fees -> The fees that need to be paid for the stuff that aren't inputs
+ */
+
+static const int TOTAL_TRIES = 100000;
+
+bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret, CAmount not_input_fees)
+{
+    out_set.clear();
+    CAmount value_track = 0;
+
+    int depth = 0;
+    std::vector<bool> selection; // select the utxo at this index
+    selection.assign(utxo_pool.size(), false);
+    bool backtrack = false;
+    CAmount actual_target = not_input_fees + target_value;
+
+    // Calculate lookahead
+    CAmount lookahead = 0;
+    for (const CInputCoin& utxo : utxo_pool) {
+        lookahead += utxo.effective_value;
+    }
+    if (lookahead < actual_target) {
+        return false;
+    }
+
+    // Sort the utxo_pool
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+
+    // Best solution
+    CAmount curr_waste = 0;
+    std::vector<bool> best_selection;
+    CAmount best_waste = MAX_MONEY;
+
+    // Depth First search loop for choosing the UTXOs
+    for (int i = 0; i < TOTAL_TRIES; ++i) {
+        // Conditions for starting a backtrack
+        if (value_track + lookahead < actual_target ||                // Cannot possibly reach target with the amount remaining in the lookahead.
+            value_track > actual_target + cost_of_change ||    // Selected value is out of range, go back and try other branch
+            (curr_waste > best_waste && (utxo_pool.at(0).fee - utxo_pool.at(0).long_term_fee) > 0)) { // Don't select things which we know will be more wasteful if the waste is increasing
+            backtrack = true;
+        } else if (value_track >= actual_target) {       // Selected value is within range
+            curr_waste += (value_track - actual_target); // This is the excess value which is added to the waste for the below comparison
+            // Adding another UTXO after this check could bring the waste down if the long term fee is higher than the current fee.
+            // However we are not going to explore that because this optimization for the waste is only done when we have hit our target
+            // value. Adding any more UTXOs will be just burning the UTXO; it will go entirely to fees. Thus we aren't going to
+            // explore any more UTXOs to avoid burning money like that.
+            if (curr_waste <= best_waste) {
+                best_selection.assign(selection.begin(), selection.end());
+                best_waste = curr_waste;
+            }
+            curr_waste -= (value_track - actual_target); // Remove the excess value as we will be selecting different coins now
+            backtrack = true;
+        }
+
+        // Backtracking, moving backwards
+        if (backtrack) {
+            backtrack = false; // Reset
+            --depth;
+
+            // Walk backwards to find the first utxo which had not has its second branch traversed
```

Comment:
This comment is sort of outdated. How about:

> Walk backwards to find the last included UTXO that still needs to have its omission branch traversed.

Reply:
Done